### ISSUE TITLE
Is sigmoid function corrected computed?

### ISSUE THREAD

**Issue by bjiyxo** [gcp/leela-zero#1484]
https://github.com/gcp/leela-zero/blob/9e903eda6e93fb65375ea03127e4f62268913175/src/Network.cpp#L974

https://en.wikipedia.org/wiki/Sigmoid_function
As mentioned above, sigmoid function is (1+e^-x)^-1=( 1+tanh(x/2) ) / 2,
which is different from (1+ tanh(x)) / 2.
Is it corrected computed?

**Comment by Ttl** [comment]
This definition is also a sigmoid function. The exact formula doesn't matter that much as long as LZ and training code use the same definition. The training code uses tanh(x) and this formula agrees with it.

Here the output is also shifted to range [0, 1] while the training code doesn't have this transformation. Shifting doesn't change the search behavior since it is a linear function as long as Puct is also scaled.

**Comment by OmnipotentEntity** [comment]
There is a good reason to believe that fixing this will reduce the strength of the network.

```
leelaz-next v leelaz-sig (46/400 games)
board size: 19   komi: 7.5
              wins              black         white       avg cpu
leelaz-next     29 63.04%       16 69.57%     13 56.52%    712.52
leelaz-sig      17 36.96%       10 43.48%     7  30.43%    761.83
                                26 56.52%     20 43.48%
```

This isn't statistically significant quite yet, but as you can see the version without the fix to the sigmoid function seems to be performing pretty poorly compared to the head of the next branch.

**Comment by samecos** [comment]
Eum,in train code ,we use TanH(x) for activation function.So,if we just want change a function ,we must ensure the consistency of the function in the function definition domain. Sig(x) = [( 1+TanH(x/2) ) / 2],
train code describe it with this `h_fc3 = tf.nn.tanh(tf.add(tf.matmul(h_fc2, W_fc3), b_fc3))`(tfprocess.py in row 608)

**Comment by bjiyxo** [comment]
https://github.com/gcp/leela-zero/blob/9e903eda6e93fb65375ea03127e4f62268913175/training/tf/tfprocess.py#L527
@samecos means here from my perspective.

**Comment by gcp** [comment]
>There is a good reason to believe that fixing this

"Fixing" is the wrong word. The code is correct.

**Comment by gcp** [comment]
It's not a sigmoid function. There are no sigmoid units in the Alpha (Go) Zero network.

The output of TanH is [-1..1]. Leela Zero's search code calculates with [0..1]. The code given remaps from the first to the latter.

It looks vaguely like a sigmoid. But as you observe, it isn't one. The naming of the variable is misleading simply because the original code *did* use sigmoids and it wasn't renamed.
